sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -357,6 +361,85 @@ def finish_bundle(self):
         {None: hold})
 
 
+class _PubSubSubscriptionWrapper(object):
+  """Wrapper for garbage-collecting temporary PubSub subscriptions."""
+
+  def __init__(self, subscription, should_cleanup):
+    self.subscription = subscription
+    self.should_cleanup = should_cleanup
+
+  def __del__(self):
+    if self.should_cleanup:
+      self.subscription.delete()
+
+
+class _PubSubReadEvaluator(_TransformEvaluator):
+  """TransformEvaluator for PubSub read."""
+
+  _subscription_cache = {}
+
+  def __init__(self, evaluation_context, applied_ptransform,
+               input_committed_bundle, side_inputs, scoped_metrics_container):
+    assert not side_inputs
+    super(_PubSubReadEvaluator, self).__init__(
+        evaluation_context, applied_ptransform, input_committed_bundle,
+        side_inputs, scoped_metrics_container)
+
+    source = self._applied_ptransform.transform._source
+    should_create = not source.subscription_name
+    self._subscription = _PubSubReadEvaluator.get_subscription(
+        self._applied_ptransform, source.project, source.topic_name,
+        source.subscription_name, should_create)
+
+  @classmethod
+  def get_subscription(cls, transform, project, topic, subscription_name,
+                       should_create):
+    if transform not in cls._subscription_cache:
+      from google.cloud import pubsub
+      if should_create:
+        subscription_name = 'dataflow_%x' % random.randrange(1 << 32)
+      cls._subscription_cache[transform] = _PubSubSubscriptionWrapper(
+          pubsub.Client(project=project).topic(topic).subscription(
+              subscription_name),
+          should_create)
+      if should_create:
+        cls._subscription_cache[transform].subscription.create()
+    return cls._subscription_cache[transform].subscription
+
+  def start_bundle(self):
+    pass
+
+  def process_element(self, element):
+    pass
+
+  def _read_from_pubsub(self):
+    from google.cloud import pubsub
+    with pubsub.subscription.AutoAck(
```

Comment:
Done.  Also note that the scope of the possible error is an exception in the evaluator.  After the evaluator emits the bundle, it can be retried, assuming future support of retrying evaluation on bundles.